crates/rattler_networking/src/authentication_storage/storage.rs

```diff
@@ -1,155 +1,100 @@
 //! Storage and access of authentication information
+
+use anyhow::{anyhow, Result};
+use reqwest::IntoUrl;
 use std::{
     collections::HashMap,
-    path::Path,
-    str::FromStr,
     sync::{Arc, Mutex},
 };
+use url::Url;
 
-use keyring::Entry;
-use reqwest::{IntoUrl, Url};
-
-use super::{authentication::Authentication, fallback_storage};
+use super::{
+    authentication::Authentication,
+    backends::{file::FileStorage, keyring::KeyringAuthenticationStorage},
+    StorageBackend,
+};
 
-/// A struct that implements storage and access of authentication
-/// information
-#[derive(Clone)]
+#[derive(Debug, Clone)]
+/// This struct implements storage and access of authentication
+/// information backed by multiple storage backends
+/// (e.g. keyring and file storage)
+/// Credentials are stored and retrieved from the backends in the
+/// order they are added to the storage
 pub struct AuthenticationStorage {
-    /// The store_key needs to be unique per program as it is stored
-    /// in a global dictionary in the operating system
-    pub store_key: String,
+    backends: Vec<Arc<Box<dyn StorageBackend + Send + Sync>>>,
```

Comment:
Arc Box seems superfluous. Maybe just Arc is enough?

Reply:
We want Arc for `Clone` and need to Box to store the `dyn Trait`

Reply:
but cant you use the arc for dyn too?

Reply:
You're right, I didn't know that. Thanks!